graphix/opengraph.py

```diff
@@ -0,0 +1,162 @@
+
+from __future__ import annotations
+
+import math
+from typing import TYPE_CHECKING, NamedTuple
+
+import networkx as nx
+
+from graphix.generator import generate_from_graph
+
+if TYPE_CHECKING:
+    from collections.abc import Mapping
+
+    from graphix.pattern import Pattern
+    from graphix.pauli import Plane
+
+
+class Measurement(NamedTuple):
+    """An MBQC measurement.
+
+    :param angle: the angle of the measurement. Should be between [0, 2)
+    :param plane: the measurement plane
+    """
+
+    angle: float
+    plane: Plane
+
+    def isclose(self, other: Measurement, rel_tol=1e-09, abs_tol=0.0) -> bool:
+        """Compares if two measurements have the same plane and their angles
+        are close.
+
+        Example
+        -------
+        >>> from graphix.opengraph import Measurement
+        >>> from graphix.pauli import Plane
+        >>> Measurement(0.0, Plane.XY).isclose(Measurement(0.0, Plane.XY))
+        True
+        >>> Measurement(0.0, Plane.XY).isclose(Measurement(0.0, Plane.YZ))
+        False
+        >>> Measurement(0.1, Plane.XY).isclose(Measurement(0.0, Plane.XY))
+        False
+        """
+        return math.isclose(self.angle, other.angle, rel_tol=rel_tol, abs_tol=abs_tol) and self.plane == other.plane
+
+
+class OpenGraph:
+    """Open graph contains the graph, measurement, and input and output
+    nodes. This is the graph we wish to implement deterministically
+
+    :param inside: the underlying graph state
+    :param measurements: a dictionary whose key is the ID of a node and the
+        value is the measurement at that node
+    :param inputs: a set of IDs of the nodes that are inputs to the graph
+    :param outputs: a set of IDs of the nodes that are outputs of the graph
+
+    Example
+    -------
+    >>> import networkx as nx
+    >>> from graphix.opengraph import OpenGraph, Measurement
+    >>>
+    >>> inside_graph = nx.Graph([(0, 1), (1, 2), (2, 0)])
+    >>>
+    >>> measurements = {i: Measurement(0.5 * i, Plane.XY) for i in range(2)}
+    >>> inputs = [0]
+    >>> outputs = [2]
+    >>> og = OpenGraph(inside_graph, measurements, inputs, outputs)
+    """
+
+    __inside: nx.Graph
+    __inputs: list[int]
+    __outputs: list[int]
+    __meas: Mapping[int, Measurement]
+
+    def __eq__(self, other) -> bool:
+        """Checks the two open graphs are equal
+
+        This doesn't check they are equal up to an isomorphism"""
+
+        if not nx.utils.graphs_equal(self._inside, other._inside):
+            return False
+
+        if self.inputs != other.inputs or self.outputs != other.outputs:
+            return False
+
+        if set(self.measurements.keys()) != set(other.measurements.keys()):
+            return False
+
+        return all(m.isclose(other.measurements[node]) for node, m in self.measurements.items())
+
+    def __init__(
+        self,
+        inside: nx.Graph,
+        measurements: Mapping[int, Measurement],
+        inputs: list[int],
```

Comment:
I agree that duplication does not make sense and so it would be nicer to use a `set`. The main reason I have used a list is actually to preserve the ordering of the inputs/outputs.

The ordering of inputs/outputs is required when checking if two OpenGraphs are equal, since the unitary transformation that the graph will be different if the inputs/outputs are in a different order.
There is of course a workaround by storing the inputs/outputs as a `set` and using a seperate data structure to store the ordering which we can then use when comparing the open graphs. However I felt this was too complicated for our current usage. What do you think?

Reply:
Mmm...Then we should use `list`.
Something like `ordered_inputs` may help users recognize the subtle difference.
Also please do not forget to add duplication check and appropriate documentations!

Reply:
Okay great. I have added comments in the docstring and in the code to emphasise that the inputs/outputs are ordered.

We can't add the duplication check at the moment due to using a NamedTuple which doesn't allow us to create our own ``__init__`` method to perform validation. We discussed other options in https://github.com/TeamGraphix/graphix/pull/191#discussion_r1711041039 and decided to omit validation for now in order to maintain simplicity and postpone proper validation to future work.

Reply:
How about `dataclass`?

Reply:
It originally was a `dataclass` and we agreed to move to a `NamedTuple` because it simplified many aspects. You can see the entire discussion here https://github.com/TeamGraphix/graphix/pull/191#pullrequestreview-2229643252

Reply:
I am the one responsible for suggesting `NamedTuple` as it seemed like the simplest solution to provide immutable fields. However, I realize now that I could have also suggested using `@dataclass(frozen=True)`. The original code didn't use either `dataclass` or `NamedTuple`, but instead manually initialized private fields and provided read-only properties.

Switching to dataclasses is indeed trivial, and I've made a proposition here: https://github.com/TeamGraphix/graphix/commit/c1d371e9a34069fd1b2b99c802119eee85183c75 . Given that we already use dataclasses in other parts of the graphix project, it makes sense to maintain consistency by adopting @dataclass(frozen=True).

While data validation can be achieved with `NamedTuple` by overriding `__new__`, it would not be as concise or straightforward as using a frozen dataclass.

Reply:
c1d371e9a34069fd1b2b99c802119eee85183c75 looks great to me.
Concise and idiomatic!

Reply:
Looks good to me, I've created a commit with these changes and pushed